Omit vendor install docs from scheduling task (#36985)
kubernetes/website: content/en/docs/tasks/manage-gpus/scheduling-gpus.md

{{< feature-state state="beta" for_k8s_version="v1.10" >}}

Kubernetes includes **experimental** support for managing GPUs
(graphical processing units) across several nodes.

> This page describes how users can consume GPUs, and outlines

Reviewer: There was a man who (among other things) ate a small airplane [1], but in general users do not "consume" GPUs.

It's pods that "consume" GPUs, and cluster admin that configures deployments to request such resources.  I would suggest another PR for fixing the document terminology / language.

[1] https://en.wikipedia.org/wiki/Michel_Lotito

> This page describes how users can consume GPUs, and outlines

Reviewer 2: How about  
```suggestion
This page describes how Pods access GPUs, and outlines
```
This uses similar wording from https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/

> This page describes how users can consume GPUs, and outlines

Reviewer: To be really exact, it should say "how pod containers can access GPUs", as resource requests are per-container.  Sorry for being inaccurate earlier.

some of the limitations in the implementation.

<!-- body -->

## Using device plugins

Kubernetes implements {{< glossary_tooltip text="device plugins" term_id="device-plugin" >}}
to let Pods access specialized hardware features such as GPUs.

{{% thirdparty-content %}}

As an administrator, you have to install GPU drivers from the corresponding
hardware vendor on the nodes and run the corresponding device plugin from the
GPU vendor. Here are some links to vendors' instructions:

* [AMD](https://github.com/RadeonOpenCompute/k8s-device-plugin#deployment)
* [NVIDIA](https://github.com/NVIDIA/k8s-device-plugin#quick-start)

Once you have installed the plugin, your cluster exposes a custom schedulable resource such as `amd.com/gpu` or `nvidia.com/gpu`.

You can consume these GPUs from your containers by requesting
the custom GPU resource, the same way you request `cpu` or `memory`.
However, there are some limitations in how you specify the resource
requirements for custom devices.

> GPUs are only supposed to be specified in the `limits` section, which means:

Reviewer 3: ```suggestion
Specify GPUs within the `limits` section, and observe the following limitations:
```
I'm suggesting this to avoid the expression, "supposed to".

* You can specify GPU `limits` without specifying `requests`, because

>   Kubernetes will use the limit as the request value by default.

Reviewer 3: ```suggestion
  Kubernetes uses the limit as the request value by default.
```
Using the the [present tense](https://kubernetes.io/docs/contribute/style/style-guide/#use-present-tense).

    must be equal.
  * You cannot specify GPU `requests` without specifying `limits`.

> - Containers (and Pods) do not share GPUs. There's no overcommitting of GPUs.

Reviewer: This is not true any more either.  Nvidia supports time slicing and there are external projects for fractionalization.  Intel GPU plugin has had sharing option from start (maybe 4 years?), and  supports nowadays fractionalization.

```shell
# Label your nodes with the accelerator type they have.
kubectl label nodes node1 accelerator=example-gpu-x100
kubectl label nodes node2 accelerator=other-gpu-k915
```

> That label key `accelerator` is just an example; you can use

Reviewer 3: ```suggestion
That label key, `accelerator`, is just an example; you can use
```
or
```suggestion
That `accelerator` label key is just an example; you can use
```

```shell
# Label your nodes with the accelerator type they have.
kubectl label nodes node1 accelerator=example-gpu-x100

> kubectl label nodes node2 accelerator=other-gpu-k915

Reviewer: IMHO this part of documentation is also vendor specific, at least nowadays. NFD can be used to automatically label nodes based on what PCI devices they have.  Whereas manually labeling nodes by their purpose is not GPU specific.

Author: @eero-t what I'm trying to convey is that a cluster operator can label their nodes how they like so that Pod that require a specific kind of GPU get scheduled there. The alternative is that the scheduler guesses and perhaps eventually the Pod lands on the right kind of node by luck.

The values that a particular cluster operator might choose ought to be meaningful. If they happen to buy X100 GPUs from Example, Inc. then this labelling will work great. Same if they use k915 boards from Other AG.

The vendors are fictional but the mechanism isn't.

Reviewer: My point was that manual labeling is basic k8s feature (not related to GPUs), so it does not need to be explained here.

In the Nvidia case you can use Nvidia version of NFD, which automatically adds GPU type labels to nodes:
https://github.com/NVIDIA/gpu-feature-discovery

And with other GPUs, suitable PCI ID rule can be added to standard NFD:
https://kubernetes-sigs.github.io/node-feature-discovery/stable/advanced/customization-guide.html#available-features

That way node labels would be always up to date without manual maintenance.

Author: > My point was that manual labeling is basic k8s feature (not related to GPUs), so it does not need to be explained here.

If we skip that part, the next step would be to delete the page as there is then nothing Kubernetes-specific left to retain.

Reviewer: Please read the whole comment, not just first clause.  IMHO you should not be promoting manually labeling GPU types when k8s offers automatic device labeling add-on (node-feature-discovery), and vendors listed here provide their own GPU labeling services (AMD: node-labeller, Nvidia: gpu-feature-discovery).

Author: The purpose of this PR is: omit vendor install docs from scheduling task.

If we have a second purpose, to also revisit the docs about node labelling, then that is a different purpose. Each PR should aim to have a single purpose.

Author: > one should not be promoting manually labeling GPU types

Noted. We're not promoting one or other approach; instead, we're letting readers decide.

Reviewer: Adding the missing options needed for that user decision, belonging to another PR -- fair enough.

AMD labeller instructions are vendor specific though, so I think removing most of that (except for telling what it is and pointing to its docs) would belong to this PR?

      resources:
        limits:
          nvidia.com/gpu: 1
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        – matchExpressions:

>           – key: beta.amd.com/gpu.family.RV # Raven GPU family

Reviewer 2: Preview of updated page and this section: https://deploy-preview-36985--kubernetes-io-main-staging.netlify.app/docs/tasks/manage-gpus/scheduling-gpus/#node-labeller

nit: for a more a consistent flow of this section, the page has
`kubectl describe node cluster-node-23`
The output has a label of `beta.amd.com/gpu.family.AI=1`
Then shortly after this, we have a pod manifest with an affinity but to a different gpu family. 
Non-blocking suggestion is to match the gpu family
```suggestion
          – key: beta.amd.com/gpu.family.AI # Arctic Islands GPU family
```
